### PR TITLE
filter: aws lambda - add async invocation mode

### PR DESCRIPTION
**Pull request by marcomagdy**
Add event-based/asynchronous invocation support to the Lambda filter.

Signed-off-by: Marco Magdy <mmagdy@gmail.com>

Risk Level: Low
Testing: unit tests
Docs Changes: docs in protos
Release Notes: N/A

### CODE REVIEW

**Comment by marcomagdy** [comment]
The clang-format failure are due to the latest check added by me here https://github.com/envoyproxy/envoy/pull/10477
I'll submit another PR to fix them.

**Comment by marcomagdy** [comment]
Should be fixed with (partially) by https://github.com/envoyproxy/envoy/pull/10620

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc [hunk 1]
```diff
@@ -35,10 +35,15 @@ namespace {
...
 constexpr auto egress_gateway_metadata_key = "egress_gateway";
 
-void setLambdaHeaders(Http::RequestHeaderMap& headers, absl::string_view function_name) {
+void setLambdaHeaders(Http::RequestHeaderMap& headers, absl::string_view function_name,
+                      InvocationMode mode) {
   headers.setMethod(Http::Headers::get().MethodValues.Post);
   headers.setPath(fmt::format("/2015-03-31/functions/{}/invocations", function_name));
-  headers.setCopy(Http::LowerCaseString{"x-amz-invocation-type"}, "RequestResponse");
+  if (mode == InvocationMode::Synchronous) {
+    headers.setCopy(Http::LowerCaseString{"x-amz-invocation-type"}, "RequestResponse");
```
Comment (mattklein123) [inline source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:43]: perf nit: you can use `setReference` here if you allocate constant strings in a ConstSingleton

File: api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto [hunk 2]
```diff
@@ -20,13 +20,27 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
...
   // The ARN of the AWS Lambda to invoke when the filter is engaged
   // Must be in the following format:
   // arn:<partition>:lambda:<region>:<account-number>:function:<function-name>
   string arn = 1 [(validate.rules).string = {min_len: 1}];
 
   // Whether to transform the request (headers and body) to a JSON payload or pass it as is.
   bool payload_passthrough = 2;
+
+  // Determines the way to invoke the Lambda function.
+  InvocationMode invocation_mode = 3;
```
Comment (mattklein123) [inline api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto:43]: Add a defined only constraint here.

File: source/extensions/filters/http/aws_lambda/config.cc [hunk 3]
```diff
@@ -49,8 +64,21 @@ Router::RouteSpecificFilterConfigConstSharedPtr
 AwsLambdaFilterFactory::createRouteSpecificFilterConfigTyped(
     const envoy::extensions::filters::http::aws_lambda::v3::PerRouteConfig& proto_config,
     Server::Configuration::ServerFactoryContext&, ProtobufMessage::ValidationVisitor&) {
-  return std::make_shared<const FilterSettings>(FilterSettings{
-      proto_config.invoke_config().arn(), proto_config.invoke_config().payload_passthrough()});
+  InvocationMode invocation_mode;
+  switch (proto_config.invoke_config().invocation_mode()) {
```
Comment (mattklein123) [inline source/extensions/filters/http/aws_lambda/config.cc:68]: This is a copy of above, I would make a small helper for this.

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc (hunk 1 above)
Comment (marcomagdy) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:43]: I haven't created one.
Would it be acceptable to create a ` static Http::LowerCaseString` in this function?

File: source/extensions/filters/http/aws_lambda/config.cc (hunk 3 above)
Comment (marcomagdy) [reply source/extensions/filters/http/aws_lambda/config.cc:68]: sure thing

File: api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto (hunk 2 above)
Comment (marcomagdy) [reply api/envoy/config/filter/http/aws_lambda/v2alpha/aws_lambda.proto:43]: sounds good.

File: source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc (hunk 1 above)
Comment (mattklein123) [reply source/extensions/filters/http/aws_lambda/aws_lambda_filter.cc:43]: No, we often run into ASAN/static initializer issues. Just use ConstSingleton. See other examples, it's trivial.

**Pull request merged**